shared/lib/gcs.py

```diff
@@ -16,81 +16,107 @@
-  """
-  local_file_path = _get_local_path(bucket, filename)
-  if os.path.exists(local_file_path):
-    logging.info("Using already downloaded GCS file: %s", local_file_path)
-    return local_file_path
-  return download_file(bucket, filename, use_anonymous_client)
 
-
-def _get_local_path(bucketname: str, filename: str) -> str:
-  return os.path.join(TEMP_DIR, bucketname, filename)
+  bucket = storage_client.bucket(bucket_name)
+  blobs = bucket.list_blobs(prefix=blob_name)
+  count = 0
+  for blob in blobs:
+    if blob.name.endswith("/"):
+      continue
+    # Get the relative path to the input blob. This is used to download folder.
+    relative_path = os.path.relpath(blob.name, blob_name)
+    if relative_path == ".":
+      # In this case, the blob is a file.
+      local_file_path = local_path
+    else:
+      # In this case, the blob is a folder.
+      local_file_path = os.path.join(local_path, relative_path)
+    # Create the directory if it doesn't exist.
+    local_dir = os.path.dirname(local_file_path)
+    if not os.path.exists(local_dir):
+      os.makedirs(local_dir)
+    # Download the file.
+    blob.download_to_filename(local_file_path)
+    count += 1
+  if count == 0:
+    logging.warning("No object found from %s/%s", bucket_name, blob_name)
+    return False
+  return True
+
+
+def download_blob_by_path(gcs_path: str,
+                          local_path: str,
+                          use_anonymous_client: bool = False) -> bool:
+  """Downloads file/folder given full GCS path (i.e. gs://bucket/path/to/file)
+  to a local path.
+
+  Args:
+    gcs_path: The full GCS path (i.e. gs://bucket/path/to/file/).
+    local_path: The local path to download the blob to.
+  """
+  if not is_gcs_path(gcs_path):
+    raise ValueError(f"Invalid GCS path: {gcs_path}")
+  bucket_name, blob_name = get_path_parts(gcs_path)
+  return download_blob(bucket_name, blob_name, local_path, use_anonymous_client)
+
+
+def maybe_download(gcs_path: str,
+                   local_path_prefix='/tmp',
+                   use_anonymous_client: bool = False) -> str:
+  """Downloads file/folder from a GCS path (i.e. gs://bucket/path/to/file)
+  to a local path. If the local file/folder already exists, then do nothing.
+
+  The local path expands the gcs_path pattern under local_path_prefix.
+  For example, if local_path_prefix is '/tmp', the local path will be
+  '/tmp/bucket/path/to/file'.
+
+  Args:
+    gcs_path: The full GCS path (i.e. gs://bucket/path/to/file/).
+    local_path_prefix: The local path prefix to download the folder to.
+    use_anonymous_client: Whether to use anonymous client to download the file.
+  Returns:
+    The local path of the downloaded file/folder.
+  """
+  if not is_gcs_path(gcs_path):
+    raise ValueError(f"Invalid GCS path: {gcs_path}")
+  bucket_name, blob_name = get_path_parts(gcs_path)
+  local_path = os.path.join(local_path_prefix, bucket_name, blob_name)
+  if os.path.exists(local_path):
```

Comment:
When running locally, if the gcs_path is a directory, the below can be a useful check because after reboot, I've noticed the files in /tmp/ are deleted but directory exists.

```
  if os.path.exists(local_path) and len(os.listdir(local_path)) > 0:
    # When running locally, we may already have downloaded the path.
    # But sometimes after restart, the directories in `/tmp` become
    # empty, so ensure that's not the case.
    return local_path
```

Reply:
added check